[quant][graphmode] Quantizing traced modules

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #39919 [wip][quant] Support general op modules with inplace options
* #39925 [quant][graphmode] Support quantizing `repeat`
* #39924 [quant][graphmode] Support squeeze/unsqueeze
* #39923 [quant][graphmode] Run RemoveRedundantDequantize in the end
* #39915 [quant][graphmode] Pass debug option into insert_quant_dequant pass
* #39911 [quant][graphmode] Use quantized::batch_norm in graph mode
* #39910 [quant] add quantized::batch_norm
* #39895 [quant][graphmode] Support prim::TupleUnpack and prim::TupleConstruct
* #39892 [quant][graphmode][refactor] isScalar check
* #39867 [jit][subgraph_rewriter] Support list of filters
* **#39826 [quant][graphmode] Quantizing traced modules**

Summary:
Expanding operator test coverage to traced modules

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21991266](https://our.internmc.facebook.com/intern/diff/D21991266)

jerryzh168 committed:
[quant][graphmode] Quantizing traced modules

Summary:
Expanding operator test coverage to traced modules

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Quantizing traced modules"


Summary:
Expanding operator test coverage to traced modules

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Quantizing traced modules"


Summary:
Expanding operator test coverage to traced modules

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Quantizing traced modules"


Summary:
Expanding operator test coverage to traced modules

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Quantizing traced modules"


Summary:
Expanding operator test coverage to traced modules

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21991266](https://our.internmc.facebook.com/intern/diff/D21991266)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Quantizing traced modules"


Summary:
Expanding operator test coverage to traced modules

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21991266](https://our.internmc.facebook.com/intern/diff/D21991266)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Quantizing traced modules"


Summary:
Expanding operator test coverage to traced modules

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21991266](https://our.internmc.facebook.com/intern/diff/D21991266)

[ghstack-poisoned]

vkuzo commented on aten/src/ATen/native/quantized/cpu/qmul.cpp:
```diff
@@ -145,6 +145,27 @@ class QMulScalarOut final {
+  static Tensor run(Tensor qa, Tensor b) {
+    TORCH_CHECK(qa.qscheme() == kPerTensorAffine ||
+              qa.qscheme() == kPerTensorSymmetric,
+              "Only per tensor quantization is suuported in Mul.");
```
nit: `supported`?

vkuzo commented on torch/testing/_internal/common_quantization.py:
```diff
@@ -130,16 +130,27 @@ def wrapper(*args, **kwargs):
             fn(*args, **kwargs)
     return wrapper
 
+def get_script_module(model, tracing, data):
```
is this used?

jerryzh168 replied:
this is used in test, you probably missed the changes in `test_quantize_script.py`, it's too large so the change is not displayed by default in github..

vkuzo commented on aten/src/ATen/native/quantized/cpu/qadd.cpp:
```diff
@@ -249,6 +249,16 @@ Tensor qadd_scalar_out(Tensor qa, Scalar b, Tensor out) {
 }
 
+template <bool ReLUFused = false>
+Tensor qadd_scalar_tensor(Tensor qa, Tensor b) {
```
wondering if there is context for this...b is a tensor which represents a scalar?  Is this related to how tracing works?  Maybe we can add a comment?

jerryzh168 replied:
yeah, will add a comment. tracing will trace a Scalar as tensor

vkuzo replied:
makes sense, thx!

raghuramank100 replied:
Wondering if the comment is accurate: add and add_scalar cannot be merged as add_scalar does not require a scale and zero-point.

jerryzh168 replied:
it's a little bit confusing, I meant when we add boradcasting op, we can merge add_scalar with that op, a separate task is to merge all variations of quantized::add into `quantized::add`, since we have overloading.

vkuzo commented on aten/src/ATen/native/quantized/cpu/qadd.cpp:
```diff
@@ -249,6 +249,16 @@ Tensor qadd_scalar_out(Tensor qa, Scalar b, Tensor out) {
+}
+
+template <bool ReLUFused = false>
+Tensor qadd_scalar_tensor_out(Tensor qa, Tensor b, Tensor out) {
```
same, if this is necessary can we add a comment on the reason?

vkuzo commented on torch/csrc/jit/passes/quantization/insert_observers.cpp:
```diff
@@ -636,33 +636,52 @@ graph(%self, %input, %batchnorm):
-
   const PatternInfo mul_nn_relu = PatternInfo::parse_from_str(R"(
-graph(%self, %a, %b):
+graph(%self, %a, %b, %relu):
```
more for my own understanding, this is not a logic change and just moving `%relu` out, right?

jerryzh168 replied:
this is a logic change, we are improving the support here to be able to match relu modules used multiple times and the matching in `is_relu_module` is also more accurate than `match::module["ReLU"]`

vkuzo reviewed (approved):
accepting to unblock.  Would be great if we added some comments inline on the why behind these changes, i.e. what specifically is different between the graphs generated by scripting vs tracing.

jerryzh168 committed:
Update on "[quant][graphmode] Quantizing traced modules"


Summary:
Expanding operator test coverage to traced modules

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21991266](https://our.internmc.facebook.com/intern/diff/D21991266)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Quantizing traced modules"


Summary:
Expanding operator test coverage to traced modules

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21991266](https://our.internmc.facebook.com/intern/diff/D21991266)

[ghstack-poisoned]

raghuramank100 commented on test/quantization/test_quantize_script.py:
```diff
@@ -1233,80 +1232,42 @@ def _test_op_impl(self, module, data, quantized_op, tracing=False, debug=False,
             def __init__(self, dim, inplace):
                 super(ConvNdRelu, self).__init__()
-                self.conv = conv_module[dim](1, 4, 2, 3).float()
+                self.conv = conv_module[dim](3, 3, 3).float()
```
Why is the dimension changed here?

jerryzh168 replied:
you mean the argument? they are changed because I changed the input dimensions

jerryzh168 committed:
Update on "[quant][graphmode] Quantizing traced modules"


Summary:
Expanding operator test coverage to traced modules

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21991266](https://our.internmc.facebook.com/intern/diff/D21991266)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Quantizing traced modules"


Summary:
Expanding operator test coverage to traced modules

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21991266](https://our.internmc.facebook.com/intern/diff/D21991266)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Quantizing traced modules"


Summary:
Expanding operator test coverage to traced modules

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21991266](https://our.internmc.facebook.com/intern/diff/D21991266)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Quantizing traced modules"


Summary:
Expanding operator test coverage to traced modules

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21991266](https://our.internmc.facebook.com/intern/diff/D21991266)

[ghstack-poisoned]
